src/main/java/org/sagebionetworks/web/client/widget/entity/controller/StorageLocationWidget.java

```diff
@@ -0,0 +1,193 @@
+
+	StorageLocationWidgetView view;
+	SynapseClientAsync synapseClient;
+	SynapseAlert synAlert;
+	EntityUpdatedHandler entityUpdatedHandler;
+	EntityBundle entityBundle;
+	
+	@Inject
+	public StorageLocationWidget(StorageLocationWidgetView view,
+			SynapseClientAsync synapseClient, SynapseAlert synAlert) {
+		this.view = view;
+		this.synapseClient = synapseClient;
+		this.synAlert = synAlert;
+		view.setSynAlertWidget(synAlert);
+		view.setPresenter(this);
+	}
+	
+	@Override
+	public void configure(EntityBundle entityBundle, EntityUpdatedHandler entityUpdatedHandler) {
+		this.entityBundle = entityBundle;
+		this.entityUpdatedHandler = entityUpdatedHandler;
+		clear();
+		getMyLocationSettingBanners();
+	}
+	
+	public void getMyLocationSettingBanners() {
+		synapseClient.getMyLocationSettingBanners(new AsyncCallback<List<String>>() {
+			@Override
+			public void onFailure(Throwable caught) {
+				hide();
+				view.showErrorMessage(caught.getMessage());
+			}
+			public void onSuccess(List<String> banners) {
+				view.setBannerDropdownVisible(!banners.isEmpty());
+				view.setBannerSuggestions(banners);
+				getStorageLocationSetting();
+			};
+		});
+	}
+	
+	public void getStorageLocationSetting() {
+		Entity entity = entityBundle.getEntity();
+		synapseClient.getStorageLocationSetting(entity.getId(), new AsyncCallback<StorageLocationSetting>() {
+			@Override
+			public void onFailure(Throwable caught) {
+				hide();
+				view.showErrorMessage(caught.getMessage());
+			}
+			
+			@Override
+			public void onSuccess(StorageLocationSetting location) {
+				//if null, then still show the default UI
+				if (location != null) {
+					//set up the view
+					if (location instanceof ExternalS3StorageLocationSetting) {
+						ExternalS3StorageLocationSetting setting = (ExternalS3StorageLocationSetting) location;
+						view.setBaseKey(setting.getBaseKey().trim());
+						view.setBucket(setting.getBucket().trim());
+						view.setExternalS3Banner(setting.getBanner().trim());
+						view.selectExternalS3Storage();
+					} else if (location instanceof ExternalStorageLocationSetting) {
+						ExternalStorageLocationSetting setting= (ExternalStorageLocationSetting) location;
+						view.setSFTPUrl(setting.getUrl().trim());
+						view.setSFTPBanner(setting.getBanner().trim());
+						view.selectSFTPStorage();
+					}
+				}
+			}
+		});
+	}
+	
+	@Override
+	public Widget asWidget() {
+		return view.asWidget();
+	}
+	
+	public void show() {
+		view.show();
+	}
+	
+	public void hide() {
+		view.hide();
+	}
+	
+	@Override
+	public void clear() {
+		synAlert.clear();
+		view.clear();
+	}
+
+	@Override
+	public void onSave() {
+		synAlert.clear();
+		StorageLocationSetting setting = getStorageLocationSettingFromView();
+		String error = validate(setting);
+		if (error != null) {
+			synAlert.showError(error);
+		} else {
+			//look for duplicate storage location in existing settings
```

Comment:
Good catch. I ended up doing this servlet side instead, see SynapseClientImpl.createStorageLocationSetting()